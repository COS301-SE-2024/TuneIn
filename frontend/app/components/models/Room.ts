// src/types/RoomCard.ts
export interface Room {
    backgroundImage: string;
    name: string;
    songName: string;
    artistName: string;
    description: string;
    userProfile?: string;
    username?: string;
    tags: string[];
  }
  