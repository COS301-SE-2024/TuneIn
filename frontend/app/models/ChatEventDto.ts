import { LiveChatMessageDto } from "../../api-client";

export class ChatEventDto {
	date_created?: Date;
	body?: LiveChatMessageDto;
	userID: string | null;
	errorMessage?: string;
}
