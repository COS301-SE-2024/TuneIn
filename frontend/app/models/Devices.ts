export interface Devices {
	id: string;
	is_active: boolean;
	is_private_session: boolean;
	name: string;
	type: string;
	supports_volume: boolean;
	volume_percant: number;
}
