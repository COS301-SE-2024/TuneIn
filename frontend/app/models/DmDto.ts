import { UserDto } from "../../api-client";

export type DirectMessageDto = {
	index: number;
	messageBody: string;
	sender: UserDto;
	recipient: UserDto;
	dateSent: Date;
	dateRead: Date;
	isRead: boolean;
	pID: string;
};
