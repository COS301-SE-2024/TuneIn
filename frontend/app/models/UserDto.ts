import { SongInfoDto } from "./SongInfoDto";
import { RoomDto } from "./RoomDto";

export type UserDto = {
	profile_name: string;
	userID: string;
	username: string;
	profile_picture_url: string;
	followers: {
		count: number;
		data: UserDto[];
	};
	following: {
		count: number;
		data: UserDto[];
	};
	links: {
		count: number;
		data: string[];
	};
	bio: string;
	current_song: SongInfoDto;
	fav_genres: {
		count: number;
		data: string[];
	};
	fav_songs: {
		count: number;
		data: SongInfoDto[];
	};
	fav_rooms: {
		count: number;
		data: RoomDto[];
	};
	recent_rooms: {
		count: number;
		data: RoomDto[];
	};
	friendship?: {
		status: boolean;
		accept_url: string;
		reject_url: string;
	};
};
