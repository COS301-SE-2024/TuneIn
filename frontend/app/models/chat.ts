export interface Chat {
    id: string;
    name: string;
    lastMessage: string;
    avatar: string;
  }
  