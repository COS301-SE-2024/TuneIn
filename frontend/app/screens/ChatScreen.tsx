import React, { useState } from 'react';
import { View, Text, TextInput, FlatList, TouchableOpacity, Image, StyleSheet } from 'react-native';
import { useRouter, useLocalSearchParams } from 'expo-router';
import { Ionicons } from '@expo/vector-icons';

const messages = [
  { id: '1', text: 'Hey there!', sender: 'John Doe', me: false },
  { id: '2', text: 'Hi! How are you?', sender: 'Me', me: true },
  // Add more dummy messages
];

const ChatScreen = () => {
  const [message, setMessage] = useState('');
  const { name } = useLocalSearchParams(); // Remove avatar from useLocalSearchParams
  const router = useRouter();

  const avatarUrl = 'https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg'; // Direct URL for avatar

  const handleSend = () => {
    if (message.trim()) {
      messages.push({ id: String(messages.length + 1), text: message, sender: 'Me', me: true });
      setMessage('');
    }
  };

  return (
    <View style={styles.container}>
      <View style={styles.header}>
        <TouchableOpacity onPress={() => router.back()} style={styles.backButton}>
          <Ionicons name="chevron-back" size={24} color="black" />
        </TouchableOpacity>
        <Image source={{ uri: avatarUrl }} style={styles.avatar} /> 
        <Text style={styles.headerTitle}>{name}</Text>
      </View>
      <FlatList
        data={messages}
        keyExtractor={item => item.id}
        renderItem={({ item }) => (
          <View style={[
            styles.messageContainer,
            item.me ? styles.messageContainerMe : styles.messageContainerOther
          ]}>
            {!item.me && (
              <Image source={{ uri: avatarUrl }} style={styles.messageAvatar} /> 
            )}
            <View style={[
              styles.messageBubble,
              item.me ? styles.messageBubbleMe : styles.messageBubbleOther
            ]}>
              <Text style={styles.messageText}>{item.text}</Text>
            </View>
          </View>
        )}
        contentContainerStyle={styles.messagesContainer}
      />
      <View style={styles.inputContainer}>
        <TextInput
          placeholder="Message..."
          style={styles.input}
          value={message}
          onChangeText={setMessage}
        />
        <TouchableOpacity style={styles.sendButton} onPress={handleSend}>
          <Image source={{ uri: 'https://img.icons8.com/material-outlined/24/000000/filled-sent.png' }} style={styles.sendIcon} />
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFFFFF',
    paddingHorizontal: 10,
  },
  header: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 15,
    paddingHorizontal: 10,
    borderBottomWidth: 1,
    borderBottomColor: '#E0E0E0',
  },
  backButton: {
    marginRight: 16,
  },
  avatar: {
    width: 40,
    height: 40,
    borderRadius: 20,
  },
  headerTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    marginLeft: 10,
  },
  messagesContainer: {
    paddingVertical: 10,
  },
  messageContainer: {
    flexDirection: 'row',
    alignItems: 'flex-end',
    marginVertical: 4,
  },
  messageContainerMe: {
    justifyContent: 'flex-end',
  },
  messageContainerOther: {
    justifyContent: 'flex-start',
  },
  messageAvatar: {
    width: 30,
    height: 30,
    borderRadius: 15,
    marginRight: 10,
  },
  messageBubble: {
    paddingVertical: 10,
    paddingHorizontal: 15,
    borderRadius: 25,
    maxWidth: '75%',
  },
  messageBubbleMe: {
    backgroundColor: '#08bdbd', // Updated to verdigris color
    alignSelf: 'flex-end',
  },
  messageBubbleOther: {
    backgroundColor: '#ECECEC',
    alignSelf: 'flex-start',
  },
  messageText: {
    fontSize: 16,
  },
  inputContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 5,
    paddingHorizontal: 10,
    borderTopWidth: 1,
    borderTopColor: '#E0E0E0',
    backgroundColor: '#FFFFFF',
  },
  input: {
    flex: 1,
    backgroundColor: '#F0F0F0',
    paddingVertical: 10,
    paddingHorizontal: 15,
    borderRadius: 25,
    marginRight: 20,
    marginBottom: 10,
    marginTop: 10,
    marginLeft: 10,
  },
  sendButton: {
    padding: 10,
  },
  sendIcon: {
    width: 24,
    height: 24,
  },
});

export default ChatScreen;
