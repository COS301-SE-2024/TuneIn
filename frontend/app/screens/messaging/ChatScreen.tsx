import React, { useEffect, useState, useRef } from "react";
import {
	View,
	Text,
	TextInput,
	FlatList,
	TouchableOpacity,
	Image,
	StyleSheet,
	ToastAndroid,
} from "react-native";
import { useRouter, useLocalSearchParams } from "expo-router";
import { Ionicons } from "@expo/vector-icons";
import MessageItem from "../../components/MessageItem";
import auth from "../../services/AuthManagement";
import * as utils from "../../services/Utils";
import { useLive } from "../../LiveContext";
import { DirectMessage } from "../../hooks/useDMControls";
import axios from "axios";
import { colors } from "../../styles/colors";
import Feather from "@expo/vector-icons/Feather";
import { UserDto } from "../../../api";

const ChatScreen = () => {
	const {
		currentUser,
		enterDM,
		leaveDM,
		dmControls,
		dmParticipants,
		directMessages,
		socketHandshakes,
	} = useLive();
	const [message, setMessage] = useState<string>("");
	const [dmError, setError] = useState<boolean>(false);
	const [unreadCount, setUnreadCount] = useState<number>(0); // To store unread message count
	const [showBanner, setShowBanner] = useState<boolean>(false); // Control banner visibility
	const router = useRouter();
	let { username } = useLocalSearchParams<{ username: string }>();
	const flatListRef = useRef<FlatList>(null); // FlatList reference

	const cleanup = async () => {
		console.log("Cleaning up DM");
		if (socketHandshakes.dmJoined) {
			console.log("Leaving DM");
			leaveDM();
		}
	};

	useEffect(() => {
		if (socketHandshakes.dmJoined) {
			if (!directMessages || directMessages.length === 0) {
				if (!socketHandshakes.dmsReceived) {
					dmControls.requestDirectMessageHistory();
				}
			}
		}
		// Scroll to the bottom whenever messages change
		if (directMessages.length > 0) {
			flatListRef.current?.scrollToEnd({ animated: true });
		}
	}, [messages, u]);

	// Check unread messages and show banner
	useEffect(() => {
		const unreadMessagesFromOther = messages.filter(
			(msg) => !msg.message.isRead && !msg.me,
		);
		setUnreadCount(unreadMessagesFromOther.length);
		setShowBanner(unreadMessagesFromOther.length > 0); // Show banner only if there are unread messages from the other person

		if (unreadMessagesFromOther.length > 0) {
			const oldestUnreadMessageIndex = messages.findIndex(
				(msg) => !msg.message.isRead && !msg.me,
			);
			if (oldestUnreadMessageIndex !== -1) {
				flatListRef.current?.scrollToIndex({
					index: oldestUnreadMessageIndex,
					animated: true,
				});
			}
		}
	}, [messages]);

	// on component mount
	useEffect(() => {
		const initialize = async () => {
			try {
				if (!socketHandshakes.dmJoined) {
					const user = dmParticipants.find((u) => u.username === username);
					if (!user) {
						await enterDM([username]);
						dmControls.requestDirectMessageHistory();
					}
				}
				// if (!socketHandshakes.dmsReceived) {
				// 	dmControls.requestDirectMessageHistory();
				// }
			} catch (error) {
				console.error("Failed to setup DM", error);
			}
		};
		initialize();
		return () => {
			cleanup()
				.then(() => {
					console.log("Cleaned up DM");
				})
				.catch((error) => {
					console.error("Failed to clean up DM", error);
				});
		};
	}, []);

	const handleSend = () => {
		if (!currentUser || dmParticipants.length <= 0) return;
		const newMessage: DirectMessage = {
			message: {
				index: directMessages.length,
				messageBody: message,
				sender: currentUser,
				recipient: dmParticipants[0],
				dateSent: new Date().toISOString(),
				dateRead: new Date(0).toISOString(),
				isRead: false,
				pID: "",
				bodyIsRoomID: false,
			},
			me: true,
			messageSent: false,
			isOptimistic: true,
		};
		dmControls.sendDirectMessage(newMessage);
		setMessage("");
	};

	return (
		<View style={styles.container}>
			<View style={styles.header}>
				<TouchableOpacity
					onPress={() => {
						router.back();
						cleanup()
							.then(() => {
								console.log("Cleaned up DM");
							})
							.catch((error) => {
								console.error("Failed to clean up DM", error);
							});
					}}
					testID="backButton"
					style={styles.backButton}
				>
					<Ionicons name="chevron-back" size={24} color="black" />
				</TouchableOpacity>
				{dmParticipants[0] && dmParticipants[0].profile_picture_url && (
					<Image
						source={{
							uri: dmParticipants[0].profile_picture_url,
						}}
						style={styles.avatar}
					/>
				)}
				<Text style={styles.headerTitle}>
					{dmError ? "Failed" : dmParticipants[0]?.profile_name || "Loading..."}
				</Text>
			</View>

			<FlatList
				ref={flatListRef}
				data={messages}
				keyExtractor={(item) => item.message.index.toString()}
				renderItem={({ item }) => (
					<>
						{showBanner &&
							!item.me &&
							item.message.index ===
								messages.findIndex((msg) => !msg.message.isRead && !msg.me) && (
								<View style={styles.bannerContainer}>
									<View style={styles.unreadBanner}>
										<Text style={styles.unreadBannerText}>
											You have {unreadCount} unread message
											{unreadCount > 1 ? "s" : ""}
										</Text>
									</View>
								</View>
							)}
						<MessageItem message={item} />
					</>
				)}
				contentContainerStyle={styles.messagesContainer}
				getItemLayout={(data, index) => ({
					length: 60,
					offset: 60 * index,
					index,
				})}
				onScrollToIndexFailed={(info) => {
					const wait = new Promise((resolve) => setTimeout(resolve, 500));
					wait.then(() => {
						flatListRef.current?.scrollToIndex({
							index: info.index,
							animated: true,
						});
					});
				}}
			/>

			{!dmError && (
				<View style={styles.inputContainer}>
					<TextInput
						placeholder="Message..."
						style={styles.input}
						value={message}
						onChangeText={setMessage}
						onSubmitEditing={handleSend}
						testID="messageInput"
					/>
					<TouchableOpacity
						style={styles.sendButton}
						testID="sendButton"
						onPress={handleSend}
					>
						<Feather name="send" size={24} color="black" />
					</TouchableOpacity>
				</View>
			)}
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: "#FFFFFF",
		paddingHorizontal: 10,
	},
	header: {
		flexDirection: "row",
		alignItems: "center",
		paddingVertical: 15,
		paddingHorizontal: 10,
		borderBottomWidth: 1,
		borderBottomColor: "#E0E0E0",
	},
	backButton: {
		marginRight: 16,
	},
	avatar: {
		width: 40,
		height: 40,
		borderRadius: 20,
	},
	headerTitle: {
		fontSize: 18,
		fontWeight: "bold",
		marginLeft: 10,
	},
	messagesContainer: {
		paddingVertical: 10,
		marginHorizontal: 20,
	},
	bannerContainer: {
		alignItems: "center", // Center banner and lines together
		marginVertical: 15, // Space around the banner container
		backgroundColor: "#E0E0E0",
		paddingVertical: 1,
		marginHorizontal: -20,
	},
	unreadBanner: {
		backgroundColor: colors.primary, // White background for the bubble
		borderRadius: 25, // Rounded bubble shape
		borderWidth: 2, // Grey border
		borderColor: "#E0E0E0", // Light grey color for the edges
		paddingVertical: 10, // Padding inside the bubble
		paddingHorizontal: 20, // Padding to make the bubble wider
	},
	unreadBannerText: {
		color: "#333333", // Darker text color
		fontSize: 15, // Adjust font size
		fontWeight: "bold", // Bold text
		textAlign: "center", // Center text inside the bubble
	},
	inputContainer: {
		flexDirection: "row",
		alignItems: "center",
		borderTopWidth: 1,
		borderTopColor: "#E0E0E0",
		paddingHorizontal: 10,
		paddingVertical: 5,
	},
	input: {
		flex: 1,
		height: 40,
		borderWidth: 1,
		borderColor: "#E0E0E0",
		borderRadius: 20,
		paddingHorizontal: 10,
		marginRight: 10,
	},
	sendButton: {
		padding: 10,
	},
});

export default ChatScreen;
