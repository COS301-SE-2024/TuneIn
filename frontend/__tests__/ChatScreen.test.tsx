import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import ChatScreen from "../app/screens/messaging/ChatScreen";
import { useRouter, useLocalSearchParams } from "expo-router";
// import { TextInput } from "react-native";
import { UserDto } from "../app/models/UserDto";
// import { DirectMessageDto } from "../app/models/DmDto";
import axios from "axios";

jest.mock("expo-font", () => ({
	...jest.requireActual("expo-font"),
	loadAsync: jest.fn(),
}));

jest.mock("expo-asset", () => ({
	...jest.requireActual("expo-asset"),
	fromModule: jest.fn(() => ({
		downloadAsync: jest.fn(),
		uri: "mock-uri",
	})),
}));

jest.mock("expo-router", () => ({
	useRouter: jest.fn(),
	useLocalSearchParams: jest.fn(),
}));

const mockSelf: UserDto = {
	userID: "1",
	username: "john_doe",
	profile_name: "John Doe",
	profile_picture_url:
		"https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1",
	followers: {
		count: 0,
		data: [],
	},
	following: {
		count: 0,
		data: [],
	},
	links: {
		count: 0,
		data: [],
	},
	bio: "",
	current_song: {
		title: "Song Name",
		artists: ["Artist Name"],
		cover: "",
		start_time: new Date(),
	},
	fav_genres: {
		count: 0,
		data: [],
	},
	fav_songs: {
		count: 0,
		data: [],
	},
	fav_rooms: {
		count: 0,
		data: [],
	},
	recent_rooms: {
		count: 0,
		data: [],
	},
};

const mockOtherUser: UserDto = {
	profile_name: "Jane Smith",
	userID: "2",
	username: "jane_smith",
	profile_picture_url: "",
	followers: {
		count: 0,
		data: [],
	},
	following: {
		count: 0,
		data: [],
	},
	links: {
		count: 0,
		data: [],
	},
	bio: "",
	current_song: {
		title: "",
		artists: [],
		cover: "",
		start_time: new Date(),
	},
	fav_genres: {
		count: 0,
		data: [],
	},
	fav_songs: {
		count: 0,
		data: [],
	},
	fav_rooms: {
		count: 0,
		data: [],
	},
	recent_rooms: {
		count: 0,
		data: [],
	},
	friendship: undefined,
};

// const initialMessages: DirectMessage[] = [
// 	{
// 		message: {
// 			index: 0,
// 			messageBody: "Hey there!",
// 			sender: mockSelf,
// 			recipient: mockOtherUser,
// 			dateSent: new Date(),
// 			dateRead: new Date(0),
// 			isRead: false,
// 			pID: "0",
// 		},
// 		me: true,
// 		messageSent: true,
// 	},
// 	{
// 		message: {
// 			index: 1,
// 			messageBody: "What's up?",
// 			sender: mockOtherUser,
// 			recipient: mockSelf,
// 			dateSent: new Date(),
// 			dateRead: new Date(0),
// 			isRead: false,
// 			pID: "1",
// 		},
// 		me: false,
// 		messageSent: true,
// 	},
// 	// Add more dummy messages
// ];

jest.mock("axios");
(axios.get as jest.Mock).mockImplementation((url: string) => {
	if (url.includes(":3000/users/")) {
		return Promise.resolve({ data: mockOtherUser });
	} else if (url.endsWith(":3000/users")) {
		return Promise.resolve({ data: mockSelf });
	} else if (url.endsWith(":3000/auth/spotify/tokens")) {
		return Promise.resolve({
			data: {
				access_token: "mock-access-token",
				token_type: "Bearer",
				scope: "user-read-private user-read-email",
				expires_in: 3600,
				refresh_token: "mock-refresh-token",
			},
		});
	}
	return Promise.reject(new Error("Invalid URL"));
});

global.fetch = jest.fn(() => Promise.resolve(new Response()));
jest.mock("@spotify/web-api-ts-sdk", () => ({
	SpotifyApi: {
		withAccessToken: jest.fn().mockImplementation(() => ({
			currentUser: {
				profile: jest.fn(),
			},
			player: {
				getAvailableDevices: jest.fn(),
				getPlaybackState: jest.fn(),
			},
		})),
	},
}));

describe("ChatScreen", () => {
	// const mockSetState = jest.fn();
	const mockRouter = {
		back: jest.fn(),
	};

	beforeEach(() => {
		// Reset mocks before each test
		jest.clearAllMocks();

		(useRouter as jest.Mock).mockReturnValue(mockRouter);
		(useLocalSearchParams as jest.Mock).mockReturnValue({
			friend: "jane_smith",
		});
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it("renders the chat screen correctly", () => {
		const { toJSON } = render(<ChatScreen />);
		expect(toJSON()).toMatchSnapshot();
	});

	/*
	it("sends a message correctly", () => {
		const { getByTestId, getByText } = render(<ChatScreen />);
		const messageInput = getByTestId("messageInput");

		// Assuming the send button has a specific testID
		const sendButton = getByTestId("sendButton");
		fireEvent.changeText(messageInput, "Hello");
		//expect(messageInput.props.value).toBe("Hello");
		fireEvent.press(sendButton);

		expect(messageInput.props.value).toBe("");
		const messageText = getByText("Hello");
		expect(messageText).toBeTruthy();
	});
	*/

	it("navigates back when back button is pressed", () => {
		const { getByTestId } = render(<ChatScreen />);

		// Assuming the back button has a specific testID
		const backButton = getByTestId("backButton");

		fireEvent.press(backButton);

		expect(mockRouter.back).toHaveBeenCalled();
	});
});
