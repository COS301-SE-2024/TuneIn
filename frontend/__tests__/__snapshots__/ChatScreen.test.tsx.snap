// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatScreen renders the chat screen correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingHorizontal": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#E0E0E0",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "paddingHorizontal": 10,
        "paddingVertical": 15,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "marginRight": 16,
          "opacity": 1,
        }
      }
      testID="backButton"
    >
      <Text />
    </View>
    <Image
      source={
        {
          "uri": "https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg",
        }
      }
      style={
        {
          "borderRadius": 20,
          "height": 40,
          "width": 40,
        }
      }
    />
    <Text
      style={
        {
          "fontSize": 18,
          "fontWeight": "bold",
          "marginLeft": 10,
        }
      }
    >
      John Doe
    </Text>
  </View>
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingVertical": 10,
      }
    }
    data={
      [
        {
          "id": "1",
          "me": false,
          "sender": "John Doe",
          "text": "Hey there!",
        },
        {
          "id": "2",
          "me": true,
          "sender": "Me",
          "text": "Hi! How are you?",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            [
              {
                "alignItems": "flex-end",
                "flexDirection": "row",
                "marginVertical": 4,
              },
              {
                "justifyContent": "flex-start",
              },
            ]
          }
        >
          <Image
            source={
              {
                "uri": "https://images.pexels.com/photos/3792581/pexels-photo-3792581.jpeg",
              }
            }
            style={
              {
                "borderRadius": 15,
                "height": 30,
                "marginRight": 10,
                "width": 30,
              }
            }
          />
          <View
            style={
              [
                {
                  "borderRadius": 25,
                  "elevation": 2,
                  "maxWidth": "75%",
                  "paddingHorizontal": 15,
                  "paddingVertical": 10,
                },
                {
                  "alignSelf": "flex-start",
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#ECECEC",
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              Hey there!
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            [
              {
                "alignItems": "flex-end",
                "flexDirection": "row",
                "marginVertical": 4,
              },
              {
                "justifyContent": "flex-end",
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "borderRadius": 25,
                  "elevation": 2,
                  "maxWidth": "75%",
                  "paddingHorizontal": 15,
                  "paddingVertical": 10,
                },
                {
                  "alignSelf": "flex-end",
                  "backgroundColor": "#08bdbd",
                },
              ]
            }
          >
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              Hi! How are you?
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderTopColor": "#E0E0E0",
        "borderTopWidth": 1,
        "flexDirection": "row",
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Message..."
      style={
        {
          "backgroundColor": "#F0F0F0",
          "borderRadius": 25,
          "flex": 1,
          "marginBottom": 10,
          "marginLeft": 10,
          "marginRight": 20,
          "marginTop": 10,
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 10,
        }
      }
      testID="sendButton"
    >
      <Image
        source={
          {
            "uri": "https://img.icons8.com/material-outlined/24/000000/filled-sent.png",
          }
        }
        style={
          {
            "height": 24,
            "width": 24,
          }
        }
      />
    </View>
  </View>
</View>
`;
